utilities/generateBrandColors.js

```diff
@@ -0,0 +1,91 @@
+    var textOnPrimary = ""
+    var textOnSecondary = ""
+
+    function checkContrast() {
+        const primaryContrast = chroma.contrast(textColor, primaryColor);
+        const secondaryContrast = chroma.contrast(textColor, secondaryColor);
+
+        if (primaryContrast > contrastTarget) {
+            textOnPrimary = textColor
+        } else {
+            textOnPrimary = textReverse
+        }
+
+        if (secondaryContrast > contrastTarget) {
+            textOnSecondary = textColor
+        } else {
+            textOnSecondary = textReverse
+        }
+    };
+    
+    checkContrast();
+
+    function updateCSS() {
+        root.style.setProperty(`--xpl-color-primary50`, primary50);
+        root.style.setProperty(`--xpl-color-primary100`, primary100);
+        root.style.setProperty(`--xpl-color-primary100`, primary100);
+        root.style.setProperty(`--xpl-color-primary200`, primary200);
+        root.style.setProperty(`--xpl-color-primary300`, primary300);
+        root.style.setProperty(`--xpl-color-primary400`, primary400);
+        root.style.setProperty(`--xpl-color-primary500`, primary500);
+        root.style.setProperty(`--xpl-color-primary600`, primary600);
+        root.style.setProperty(`--xpl-color-primary700`, primary700);
+        root.style.setProperty(`--xpl-color-primary800`, primary800);
+        root.style.setProperty(`--xpl-color-primary900`, primary900);
+        root.style.setProperty(`--xpl-color-primary1000`, primary1000);
+        root.style.setProperty(`--xpl-color-brand-primary-gradient-start-dm`, primaryDMGradStart);
+        root.style.setProperty(`--xpl-color-brand-primary-gradient-start-dm`, primaryDMGradEnd);
+        root.style.setProperty(`--xpl-color-text-on-primary-lm`, textOnPrimary);
+
+        root.style.setProperty(`--xpl-color-secondary50`, secondary50);
+        root.style.setProperty(`--xpl-color-secondary100`, secondary100);
+        root.style.setProperty(`--xpl-color-secondary100`, secondary100);
+        root.style.setProperty(`--xpl-color-secondary200`, secondary200);
+        root.style.setProperty(`--xpl-color-secondary300`, secondary300);
+        root.style.setProperty(`--xpl-color-secondary400`, secondary400);
+        root.style.setProperty(`--xpl-color-secondary500`, secondary500);
+        root.style.setProperty(`--xpl-color-secondary600`, secondary600);
+        root.style.setProperty(`--xpl-color-secondary700`, secondary700);
+        root.style.setProperty(`--xpl-color-secondary800`, secondary800);
+        root.style.setProperty(`--xpl-color-secondary900`, secondary900);
+        root.style.setProperty(`--xpl-color-secondary1000`, secondary1000);
+        root.style.setProperty(`--xpl-color-brand-secondary-gradient-start-dm`, secondaryDMGradStart);
+        root.style.setProperty(`--xpl-color-brand-secondary-gradient-start-dm`, secondaryDMGradEnd);
+        root.style.setProperty(`--xpl-color-text-on-secondary-lm`, textOnSecondary);
+    }
+
+    updateCSS();
+};
```

Comment:
I already spoiled the surprise but I took a stab at DRYing these functions up. Here's where I landed:

```js
import chroma from 'chroma-js';

const { style } = document.documentElement;

function generateGradientArray(color) {
    const gradientArray = [0.9, 0.7, 0.6, 0.4, 0.2];

    const morning = gradientArray.map((val) => (
        chroma.mix(color, 'white', val, 'hsl').hex()
    ));

    const noon = chroma(color).hex();

    const evening = gradientArray.reverse().map((val) => (
        chroma.mix(color, 'black', val, 'hsl').hex()
    ));

    const gradientStart = chroma.mix(color, 'white', 0.2, 'hsl').alpha(0.2).css();
    const gradientEnd = chroma.mix(color, 'white', 0.2, 'hsl').alpha(0).css();

    return [...morning, noon, ...evening, gradientStart, gradientEnd];
}

function setTextColors(color, themeName = 'primary') {
    const textColor = '#292632';
    const textReverse = '#FFFFFF';
    const isTextColorAcceptable = chroma.contrast(textColor, color) > 4.5;

    style.setProperty(
        `--xpl-color-text-${themeName}-lm`,
        isTextColorAcceptable ? textColor : textReverse,
    );
    style.setProperty(
        `--xpl-color-text-${themeName}-dm`,
        isTextColorAcceptable ? textReverse : textColor,
    );
}

function setCustomProperties(colorSetName, gradientMode, colorSet) {
    const [gradientEnd, gradientStart, ...setArray] = colorSet.reverse();

    setArray.reverse().forEach((hexDex, idx) => {
        const int = idx === 0 ? 50 : idx * 100;
        style.setProperty(`--xpl-color-${colorSetName}${int}`, hexDex);
    });

    style.setProperty(`--xpl-color-brand-${colorSetName}-gradient-start-${gradientMode}`, gradientStart);
    style.setProperty(`--xpl-color-brand-${colorSetName}-gradient-end-${gradientMode}`, gradientEnd);
}

export {
    generateGradientArray,
    setCustomProperties,
    setTextColors,
};
```

If the color in question is `'#6923f4'`, these function set value to the following variables:

```
'--xpl-color-text-primary-lm' '#FFFFFF'
'--xpl-color-text-primary-dm' '#292632'

'--xpl-color-primary50' '#f0e9fe'
'--xpl-color-primary100' '#d2bdfc'
'--xpl-color-primary200' '#c3a7fb'
'--xpl-color-primary300' '#a57bf8'
'--xpl-color-primary400' '#874ff6'
'--xpl-color-primary500' '#6923f4'
'--xpl-color-primary600' '#4e0bd5'
'--xpl-color-primary700' '#3b089f'
'--xpl-color-primary800' '#27056a'
'--xpl-color-primary900' '#1d0450'
'--xpl-color-primary1000' '#0a011b'

'--xpl-color-brand-primary-gradient-start-lm' 'rgba(135,79,246,0.2)'
'--xpl-color-brand-primary-gradient-end-lm' 'rgba(135,79,246,0)'
```

See anything wrong with these vars man?